pkg/cloudprovider/provider/linode/provider.go

```diff
@@ -0,0 +1,447 @@
+	}
+	c.PrivateNetworking, err = p.configVarResolver.GetConfigVarBoolValue(rawConfig.PrivateNetworking)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	for _, tag := range rawConfig.Tags {
+		tagVal, err := p.configVarResolver.GetConfigVarStringValue(tag)
+		if err != nil {
+			return nil, nil, err
+		}
+		c.Tags = append(c.Tags, tagVal)
+	}
+
+	return &c, &pconfig, err
+}
+
+func (p *provider) AddDefaults(spec v1alpha1.MachineSpec) (v1alpha1.MachineSpec, error) {
+	return spec, nil
+}
+
+func (p *provider) Validate(spec v1alpha1.MachineSpec) error {
+	c, pc, err := p.getConfig(spec.ProviderSpec)
+	if err != nil {
+		return fmt.Errorf("failed to parse config: %v", err)
+	}
+
+	if c.Token == "" {
+		return errors.New("token is missing")
+	}
+
+	if c.Region == "" {
+		return errors.New("region is missing")
+	}
+
+	if c.Type == "" {
+		return errors.New("type is missing")
+	}
+
+	_, err = getSlugForOS(pc.OperatingSystem)
+	if err != nil {
+		return fmt.Errorf("invalid operating system specified %q: %v", pc.OperatingSystem, err)
+	}
+
+	ctx := context.TODO()
+	client := getClient(c.Token)
+
+	_, err = client.GetRegion(ctx, c.Region)
+	if err != nil {
+		return err
+	}
+
+	_, err = client.GetType(ctx, c.Type)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *provider) Create(machine *v1alpha1.Machine, _ *cloud.MachineCreateDeleteData, userdata string) (instance.Instance, error) {
+	c, pc, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	ctx := context.TODO()
+	client := getClient(c.Token)
+
+	sshkey, err := ssh.NewKey()
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate ssh key: %v", err)
+	}
+
+	slug, err := getSlugForOS(pc.OperatingSystem)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, invalid operating system specified %q: %v", pc.OperatingSystem, err),
+		}
+	}
+	createRequest := linodego.InstanceCreateOptions{
+		Image:          slug,
+		Label:          machine.Spec.Name,
+		Region:         c.Region,
+		Type:           c.Type,
+		PrivateIP:      c.PrivateNetworking,
+		BackupsEnabled: c.Backups,
+		AuthorizedKeys: []string{sshkey.PublicKey},
+		Tags:           append(c.Tags, string(machine.UID)),
+		StackScriptID:  cloudinitStackScriptID,
+		StackScriptData: map[string]string{
+			"userdata": base64.StdEncoding.EncodeToString([]byte(userdata)),
+		},
+	}
+
+	linode, err := client.CreateInstance(ctx, createRequest)
+	if err != nil {
+		return nil, doStatusAndErrToTerminalError(err)
+	}
+
+	//We need to wait until the linode really got created as tags will be only applied when the linode is running
+	err = wait.Poll(createCheckPeriod, createCheckTimeout, func() (done bool, err error) {
```

Comment:
If I am not misunderstating the API docs, it should be possible to pass Tags on creation: https://developers.linode.com/api/v4#operation/createLinodeInstance

That would be much nicer because with the current approach we basically lose track of the instance if the controller crashes at the right point in time

Reply:
That's what I get for borrowing from another provider. I'll take another look at this.

Reply:
updated in [575403a](https://github.com/kubermatic/machine-controller/commit/575403ae478681a8e3395eeabbb8588c522309a5)

simplified the polling using a wait mechanism built into the linodego client.